cli/command/manifest/push.go

```diff
@@ -0,0 +1,623 @@
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/distribution/registry/api/v2"
+	"github.com/docker/distribution/registry/client"
+	digest "github.com/opencontainers/go-digest"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/cli/cli/manifest/fetcher"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/docker/docker/registry"
+)
+
+type pushOpts struct {
+	newRef string
+	file   string
+	purge  bool
+}
+
+// YamlInput represents the YAML format input to the pushml
+// command.
+type YamlInput struct {
+	Image     string
+	Manifests []YamlManifestEntry
+}
+
+// YamlManifestEntry represents an entry in the list of manifests to
+// be combined into a manifest list, provided via the YAML input
+type YamlManifestEntry struct {
+	Image    string
+	Platform manifestlist.PlatformSpec
+}
+
+// we will store up a list of blobs we must ask the registry
+// to cross-mount into our target namespace
+type blobMount struct {
+	FromRepo reference.Named
+	Digest   digest.Digest
+}
+
+// if we have mounted blobs referenced from manifests from
+// outside the target repository namespace we will need to
+// push them to our target's repo as they will be references
+// from the final manifest list object we push
+type manifestPush struct {
+	Name      string
+	Digest    string
+	JSONBytes []byte
+	MediaType string
+}
+
+type manifestListPush struct {
+	targetRepoInfo    *registry.RepositoryInfo
+	targetRef         reference.Named
+	targetEndpoint    registry.APIEndpoint
+	targetName        string
+	list              manifestlist.ManifestList
+	mountRequests     []manifestPush
+	blobMountRequests []blobMount
+}
+
+func newPushListCommand(dockerCli command.Cli) *cobra.Command {
+
+	opts := pushOpts{}
+
+	cmd := &cobra.Command{
+		Use:   "push [newRef | --file pre-annotated-yaml]",
+		Short: "Push a manifest list for an image to a repository",
+		Args:  cli.RequiresMinArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return putManifestList(dockerCli, opts, args)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVarP(&opts.file, "file", "f", "", "Path to a file containing a manifest list and its annotated constituent manifests")
+	flags.BoolVarP(&opts.purge, "purge", "p", true, "After pushing, delete the user's locally-stored manifest list info")
+
+	return cmd
+}
+
+func putManifestList(dockerCli command.Cli, opts pushOpts, args []string) error {
+	var (
+		yamlInput                         YamlInput
+		initialRef                        string
+		listPush                          manifestListPush
+		fullTargetRef, targetRef, bareRef reference.Named
+		err                               error
+	)
+
+	// First get all the info we'll need from either a yaml file, or a user's locally-creatd manifest transaction.
+	if err := checkArgs(opts, args); err != nil {
```

Comment:
ExactArgs(1) won't work b/c --file foo.yaml would be a zero-arg command. But -- I will change it to MaxArgs(1) instead since the only way to really know if what the user did is right is to try and use the arg as a ref name and print the resulting error message if they tried to use a yaml file as the arg (without --file). 👍